### ISSUE TITLE
HttpClient Response Stream hangs

### ISSUE THREAD

**Issue by iSazonov** [dotnet/runtime#74568]
There have already been two similar requests #36822 and #11594.
In #36822 there is even an excellent repro https://github.com/dotnet/runtime/issues/36822#issuecomment-632821881.

I discovered the issue in PowerShell:
1. Start download large file
```powershell
Invoke-WebRequest https://github.com/PowerShell/PowerShell/releases/download/v7.3.0-preview.7/PowerShell-7.3.0-preview.7-win-x64.zip -OutFile C:\temp\PowerShell-7.3.0-preview.7-win-x64.zip
```
2. Emulate network disaster by means of two options:
2.1 Close Wi-Fi connection on notebook
2.2 Disconnect Wi-Fi router from Internet
3. PowerShell cmdlet hangs on CopyToAsync()
- for option 2.1 until Wi-Fi recovered (if network was locked over ~1 min download is interrupted)
- for option 2.2 _infinitly_ even after the Internet connection is restored

It is not a problem for interactive scenario since user can press Ctrl-C but it is an unpleasant problem for script and (PowerShell SDK custom) host scenarios.
Since it does not leave a trace in the logs and is eliminated by rebooting, the technical support concludes that it is a hardware problem, which is a false path.
PowerShell user also does not get a message that the file is not fully downloaded.

Note that it is the _web_ client that we are considering. Why is it important to have a timeout?

The infinite timeout of sockets is a fundamental constant and this will never be changed. This is useful for low-level connections which can re-establish the connection without any extra effort.
But how do _web_ services behave? In fact they close the connection if the client doesn't send requests for some time so as not to waste resources. This timeout is quite short.
But if any web server closes a session by a short timeout, then it doesn't make sense for any web client to have an infinite timeout on socket level - reconnection on the socket level doesn't work. _On the contrary, the client must receive a timeout signal to either end the session on his side or try to resume it at a higher level._
This makes us think that any web client should have a default timeout greater than the typical timeout of any web server. If there is any doubt that there should be a default timeout, then at least the developer should be able to set it.

----

In which APIs is it better to add this timeout? I don't have the exact answer.

Obviously the timeout must be in SocketsHttpHandler.

But this is not enough since this class is sealed. Thus PowerShell historically uses HttpClientHandler which is a SocketsHttpHandler wrapper. We could migrate PowerShell to SocketsHttpHandler but HttpClientHandler uses some _internal_ helper classes to map to SocketsHttpHandler. So perhaps it makes sense to add ReadWriteSocketTimeout to HttpClientHandler too (or make the helper classes public).

Note that the WebRequest API supports this timeout (although it doesn't seem to be public either).
https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1666-L1702

This uses a callback, which is not convenient. In addition, it duplicates the standard code from the SocketsHttpHandler, which can be changed in the future but will not be automatically inherited by the application.

**Comment by karelz** [comment]
Triage: There may be opportunity to provide more user-friendly timeouts for ResponseStream read/write. Currently it has to go via ConnectCallback or write your own streams, which is rather ugly.
We should figure out what's the best option.

Note: There is difference between overall timeout and idle timeout on the read/write.

**Comment by iSazonov** [comment]
SocketsHttpHandler return NetworkStream instance
https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1588

[NetworkStream docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.networkstream.readtimeout?view=net-6.0) says about ReadTimeout(WriteTimeout):
> This property affects only _synchronous_ reads performed by calling the [Read](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.networkstream.read?view=net-6.0) method. This property does not affect asynchronous reads performed by calling the [BeginRead](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.networkstream.beginread?view=net-6.0) method.

I think it is true for ReadAsync/WriteAsync too.

HttpClient doesn't use NetworkStream directly but by means of some stream wrappers. So if NetworkStream doesn't throw on socket timeout these wrappers don't too.
Then NetworkStream is a wrapper for Socket class.

So I assume Socket would be the right place to throw on socket timeout.

Since it is all about async and streams I would welcome comments from @stephentoub and @adamsitnik experts.

**Comment by stephentoub** [comment]
> I think it is true for ReadAsync/WriteAsync too.

No, it doesn't.

**Comment by iSazonov** [comment]
> > I think it is true for ReadAsync/WriteAsync too.
> 
> No, it's not.

Oh, then I don't understand why NetworkStream/Socket is not throw on timeout. Is this expected or bug?

**Comment by stephentoub** [comment]
> then I don't understand why NetworkStream/Socket is not throw on timeout. Is this expected or bug?

I don't understand the question. ReadTimeout/WriteTimeout does not apply to asynchronous operations so the async operations don't timeout in response to them. Read/WriteAsync are cancelable via the token passed in.

**Comment by iSazonov** [comment]
> I don't understand the question.

If I understand correctly socket pinvoke read operation return a timeout error code in the Issue scenario 
I guess for Windows it is [WSAETIMEDOUT](https://docs.microsoft.com/en-us/windows/desktop/WinSock/windows-sockets-error-codes-2)
So if it is not Success I'd expect an exception.

**Comment by stephentoub** [comment]
For sync operations, not async operations. You said the timeout applies to ReadAsync/WriteAsync as well, and I said it doesnt. It doesn't apply at the .NET level nor on Windows level with overlapped I/O.

**Comment by stephentoub** [comment]
Note that PowerShell can implement timeouts via cancellation per read/write operation just as well as SocketsHttpHandler could internally.

**Comment by iSazonov** [comment]
> For sync operations, not async operations. You said the timeout applies to ReadAsync/WriteAsync as well, and I said it doesnt. It doesn't apply at the .NET level nor on Windows level with overlapped I/O.

In PowerShell I migrate to SocketHttpHandler and implement callback as in HttpWebRequest.cs metioned above. Only I set ReceiveTimeout and SendTimeout for socket in async branch. I see in debbuger the socket is closed if network fails (I hope it is not debugger side effect) but CopyToAsync() is still blocked on copyTask.Wait().
The PowerShell code (without my additions for sockets)
https://github.com/PowerShell/PowerShell/blob/0c63a549528ceb8eef68cef3a441a63401a7f291/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs#L286-L309


> Note that PowerShell can implement timeouts via cancellation per read/write operation just as well as SocketsHttpHandler could internally.

I see no other way than to check for changes in the length of the target stream in a separate task. But it is a workaround, not a solution to this problem, which looks dangerous.

**Comment by stephentoub** [comment]
> but CopyToAsync() is still blocked on copyTask.Wait().

And in the debugger you see that the cancellation token passed to both CopyToAsync and Wait has had cancellation requested?

**Comment by iSazonov** [comment]
The cancellation token works well if PowerShell user press Ctrl-C. 
In experiment I do the cancellation token is not involved.

It is .Net 7.0 Preview.3
It is ContentLegthReadStream -> SslStream -> NetworkStream

Repro steps (there was an error in my previous post, sorry):
1. Start PowerShell cmdlet
2. PowerShell show progress bar is updated.
3. Disconnect network
4. Progress bar looks like freeze since stream.Position is not changed.
5. Debugger show that the `while` cycle is working infinitely because the copyTask is never moved to completion status.
If I expand `input` stream properties I see `socket` in _Closed_ status.

Since the socket is in Closed status I'd expect the task should complete and since it is unexpected socket close I'd expect an exception.

**Comment by stephentoub** [comment]
> In experiment I do the cancellation token is not involved.

It should be. If PowerShell wants to time out the operation, it has the ability to do so via the token.

**Comment by iSazonov** [comment]
I don't understand what timeout you mean. The scenario is about network disaster. If network is disconnected read operation should return an I/O error on the socket unexpected closing and the task should throw an exception.

**Comment by stephentoub** [comment]
You were saying you wanted to be able to set a read/write timeout. I'm telling you how.

If your concern is the socket connection failing at the OS level or below, it's up to the OS to tell us that. It will eventually, but it can take many seconds or even minutes depending on the OS and its configuration. There's nothing short of cancellation/timeout to be done about that at the .NET level.

**Comment by iSazonov** [comment]
I may not have been accurate, sorry, but it's not about read/write timeout operations, it's about timeout on the socket level itself and this timeout works - OS close the socket after network disaster and .Net see this. A problem is that NetworkStream continue think that the socket lives and copy task is _never_ completed.

**Comment by stephentoub** [comment]
> A problem is that NetworkStream continue think that the socket lives and copy task is never completed

That is an OS issue. If the OS isn't reporting that the connection has aborted, there's nothing NetworkStream/Socket can do about that other than user-initiated timeout/cancellation.

**Comment by stephentoub** [comment]
> But the socket is already in Closed status

What member specifically are you looking at that says "Closed"?

**Comment by iSazonov** [comment]
Connected (on srceenshot network is in up state, if network is in down state this has False value after ~1min)
![image](https://user-images.githubusercontent.com/22290914/186943591-d1e79101-93cd-4598-93de-b166a6f64d0d.png)

**Comment by iSazonov** [comment]
I think that if the OS has closed the socket, no doubt ReadAsync task should end - it has nothing to do over a closed socket. 
Since OS can not know the reason for closing the socket, an exception cannot be generated unless the connection is closed by timeout.
I think in any case it is ContentLegthReadStream to know if full content is read and generate an exception if not and I guess it does this.

**Comment by iSazonov** [comment]
Maybe problem in the method https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs,1719
Comment says "Throws IOException on EOF.  This is only called when we expect more data." but it is not throw if network is disconnected.

**Comment by iSazonov** [comment]
NetworkStream doen't overload ReadAsync() method but SslStream does. May issue is in SslStream...

**Comment by stephentoub** [comment]
> Connected (on srceenshot network is in up state, if network is in down state this has False value after ~1min)

I expect this is a case of Schrodinger's cat. You're only seeing that as Connected=false because you queried the properties of the Socket in the debugger. Also, note that this Connected property doesn't actually mean the connection has been dropped.  It's set to false if many properties, like Available, throw an exception.

> I think that if the OS has closed the socket, no doubt ReadAsync task should end

It does.  If you believe otherwise, please share an isolated repro, without PowerShell involved, that demonstrates this not happening _once the OS deems the connection has been lost_.

> but it is not throw if network is disconnected.

Yes, it does.  If the socket is closed, the OS will alert the Socket to that fact and the Read{Async} will end.

**Comment by iSazonov** [comment]
If remove using CTS in the repro it outputs infinitly:
<details>
<summary>Code</summary>
<ul>
<li>
        static async Task Main(string[] args)
        {
            var httpClient = new HttpClient();

            var request = new HttpRequestMessage(HttpMethod.Get, "https://speed.hetzner.de/10GB.bin");
            var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
            Console.WriteLine($"Content-Length: {response.Content.Headers.ContentLength.Value}");


            var responseStream = await response.Content.ReadAsStreamAsync();
            var memoryStream = new MemoryStream();

            var t = responseStream.GetType();
            var _connection = t.GetField("_connection", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic)?.GetValue(responseStream)!;
            var t_connection = _connection.GetType();
            var _stream = (System.Net.Security.SslStream)t_connection.GetField("_stream", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic)?.GetValue(_connection)!;
            var t_stream = _stream.GetType();
            var innerStream = t_stream.GetProperty("InnerStream", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic)?.GetValue(_stream)!;
            var t_innerStream = innerStream.GetType();
            var innerStream_connection = t_innerStream.GetField("_connection", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic)?.GetValue(innerStream)!;
            var t_innerStream_connection = innerStream_connection.GetType();
            var network_stream = (System.Net.Sockets.NetworkStream)t_innerStream_connection.GetField("_stream", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic)?.GetValue(innerStream_connection)!;
            var s = network_stream.Socket;
            s.ReceiveTimeout = 10; // This doesn't work
            s.SendTimeout = 10; // This doesn't work

            long previousLength = 0;
            var cts = new CancellationTokenSource();
            cts.CancelAfter(2000);

            _ = Task.Run(async () =>
            {
                while (true)
                {
                    Console.WriteLine($"[{DateTimeOffset.Now.ToString(CultureInfo.InvariantCulture)}] Memory stream length: {memoryStream.Length} Socket: Blocking = {s.Blocking} Connected = {s.Connected} SafeHandle.IsClosed = {s.SafeHandle.IsClosed} SafeHandle.IsInvalid = {s.SafeHandle.IsInvalid}");
                    if (previousLength != memoryStream.Length)
                    {
                        previousLength = memoryStream.Length;
                        cts.CancelAfter(2000);
                    }

                    await Task.Delay(1000);
                }
            });

            await responseStream.CopyToAsync(memoryStream, cts.Token);
        }
</li>
</ul>
</details>


```
[08/29/2022 15:27:44 +05:00] Memory stream length: 17579708 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:45 +05:00] Memory stream length: 20741820 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:46 +05:00] Memory stream length: 24035004 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:47 +05:00] Memory stream length: 27754172 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:48 +05:00] Memory stream length: 31506108 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:49 +05:00] Memory stream length: 33357500 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:50 +05:00] Memory stream length: 33357500 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
[08/29/2022 15:27:51 +05:00] Memory stream length: 33357500 Socket: Blocking = True Connected = True SafeHandle.IsClosed = False SafeHandle.IsInvalid = False
```

So the socket is never closed. The fact that a read operation never terminates if there is no network connectivity does not make sense. So the original request to add a timeout for each read operation is still valid.

There is a related discussion https://github.com/dotnet/runtime/issues/916#issuecomment-562083237
It doesn't give me any understanding why it's impossible to use the current CTS in ReadAsync and create a linked CTS with a timer based on it, like it is done elsewhere. 
I hope there is a way to pass the Timeout from HttpClient to NetworkStream and Socket.

**Comment by stephentoub** [comment]
> So the original request to add a timeout for each read operation is still valid.

As I've already stated, PowerShell itself can already do this via a CancellatioToken. Are you asking for this again because of convenience or because there's a technical blocker?

> why it's impossible to use the current CTS in ReadAsync and create a linked CTS with a timer based on it

It's not impossible. It's expensive. And it's exactly what PowerShell itself can do if it wants a timeout.

**Comment by iSazonov** [comment]
This is neither a convenience nor a technical interlock. It is something that is misleading and previous discussions show it. If HttpClient has a Timeout property and it doesn't work as users expect, it's really costly for them because first they have to take the time to understand the problem and then still come up with a workaround. Same thing happens with PowerShell web cmdlets - users keep asking why Timeout doesn't work sometimes.

> It's not impossible. It's expensive (relatively). And it's exactly what PowerShell itself can do if it wants a timeout.

Impossible or still expensive? This was already implemented for connect and send. Why not for the read? If we look on ContentLengthReadStream -> SslStream -> NetworkStream -> Socket - all the classes have timeout properties except ContentLegthReadStream. The properties are used for sync operations and I hope they could be reused for async too.

> And it's exactly what PowerShell itself can do if it wants a timeout.

The CTS is for _whole_ ReadAsync method but the request is to have timeout for every internal io operation. Otherwise it creates a clumsy workaround as shown in the code example above.

**Comment by MihaZupan** [comment]
For reference, YARP has logic similar to what Stephen described above: [`StreamCopier.CopyAsync`](https://github.com/microsoft/reverse-proxy/blob/595d47bb97375845475c07c1fa548211fbb7ceec/src/ReverseProxy/Forwarder/StreamCopier.cs#L71-L104).
It also includes instrumentation for Telemetry - you could choose to do something similar if your aim is to track the progress as well (even in cases where the destination stream's `Length` may not be supported).

**Comment by iSazonov** [comment]
Thanks for workarounds. I can not use them because this is ContentLengthReadStream - internal helper class in HttpConnection.  ContentLengthReadStream uses [CopyToContentLengthAsync](https://source.dot.net/System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs.html#087291662a922514) from HttpConnection and it is not trivial.

> This is as good as SocketsHttpHandler itself could do it (there are possible further tweaks to minimize overheads for this or that case, but fundamentally SocketsHttpHandler doesn't have any magic sauce and is just doing what PowerShell itself can do on top).
>  if you wanted a CopyToAsync operation that applied the same timeout to each read it issues while being efficient about reusing the CancellationTokenSource:

Yes, this is I want. And HttpClient really does such magic for some _whole_ operations - it convert Timeout property to linked CTS with the timeout. I hope there is a way to specify CTS to each read in CopyToAsync.

**Comment by stephentoub** [comment]
> . I can not use them because this is ContentLengthReadStream - internal helper class in HttpConnection. ContentLengthReadStream uses [CopyToContentLengthAsync](https://source.dot.net/System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs.html#087291662a922514) from HttpConnection and it is not trivial.

That's an implementation detail that doesn't affect your ability to use the examples posted. You do not need to copy it nor can or should you. That same Stream's ReadAsync encompasses all of the logic to make it work as well.

**Comment by iSazonov** [comment]
:-) From the point of view of an expert like you, it looks simple. I would have to throw out some of the standard code and think about what I could break. But then what is the purpose of .Net Runtime, and in particular HttpClient, if not to allow a wider range of developers to create simple and reliable code without going into the details of the implementation, which is created by more qualified experts?
From my perspective as a consumer, if HttpClient.Timeout works for all but one operation, it's a bug. It's like if the button on my floor in the lift didn't work.
There are of course many classes and dependencies, but the path from HttpClient.Timeout to ContentLengthReadStream does not look hopeless HttpClient -> SocketsHttpHandler -> HttpConnectionSettings - > HttpConnection -> ContentLengthReadStream

**Comment by stephentoub** [comment]
> From my perspective as a consumer, if HttpClient.Timeout works for all but one operation, it's a bug

Can you help me understand what about the docs for HttpClient.Timeout suggests it's a per-read/write timeout?

**Comment by iSazonov** [comment]
> Gets or sets the timespan to wait before the request times out.

> The same timeout will apply for all requests using this [HttpClient](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-6.0) instance.

You can certainly argue about what a _request_ is, but in my understanding it is _any_ request to the server, and since it comes from HttpClient, I expect that all its settings apply to all derived actions.
In other words, if the stream was created in isolation as a `new AnyWebStream(url)`, then the expectation is that it works the way AnyWebStream itself is implemented. But if some stream is created inside an HttpClient, the expectation is that Timeout acts on all its operations, including this internal stream too, not just on send or post.

Also the mention in the documentation of DNS inaccessibility looks strange in the sense that it's some third-party service while the _main_ service is inaccessible. Common sense says that the main service is what the HttpClient consumer thinks about and expects Timeout to apply to the unavailability of the main service in any HttpClient operation, including retrieving Content.

**Comment by stephentoub** [comment]
Thanks.  That didn't really answer my question, though.  I'm confused about what exactly we're discussing here...

In some parts of the conversation, it seems like your concern is that the OS is notifying the request that the socket has been closed but the operation isn't respecting that.  I believe that's been shown to not be the case; in these cases, as far as what the OS is reporting, the socket is still alive and well and we have no reason to terminate the read.

In some parts of the conversation, it seems like your concern is lack of an ability to specify a timeout _per_ Socket.ReadAsync/WriteAsync that's performed as part of the HTTP request/response (e.g. each read gets 5 seconds to complete regardless of how many previous reads have been issued and how long each of those has taken).  There is no built-in mechanism for that, but as has been exemplified numerous times, it's straightforward to layer on top of the CancellationToken at various layers of granularity.

In some parts of the conversation, it seems like your concern is that HttpClient.Timeout only applies to the operations performed on the HttpClient and not operations on a Stream it creates.  That is correct.  The timeout applies until the moment that the Task the HttpClient gives you back has completed.  After that, you control any cancellation/timeout via the CancellationToken you pass to whatever operations you subsequently call.

In some parts of the conversation, it seems like the concern is lack of detailed documentation about HttpClient.Timeout.  That's fair, and the docs can be improved.

In some parts of the conversation, it seems like the concern is that PowerShell isn't behaving well, and you're placing blame for that on HttpClient.  I believe we've shown that this is on PowerShell to fix by correctly passing in a CancellationToken with whatever timeout semantics are desired.

**Comment by iSazonov** [comment]
> I'm confused about what exactly we're discussing here...

This is a search for an answer to why HttpClient.Timeout does not work in the particular scenario as expected.
I insist that it is the expected behavior of users that this timeout should work. This certainty comes from previous discussions, it has been reinforced during this discussion. Furthermore, yesterday I posted a PR with a workaround in the PowerShell repository, and the first comment from the MSFT PowerShell team member was whether this is a bug in .Net, indicating that even the much experienced developers perceive HttpClient.Timeout just as much as less experienced ones like me.

It's not just user expectations that this timeout should work, but also the implementation itself. These streams like ContentLengthReadStream are internal HttpClient classes that are spawned _only in HttpClient_ and nowhere else. Such streams must have the same timeout. This does not contradict the basic Stream class concept, which initially supports such timeouts (abstract CanTimeout property).

As for the documentation, it simply says that this timeout exists, which supports user expectations that this timeout works for _all_ operations without exceptions. Of course the documentation can be updated that this doesn't work, but it will still raise questions and be perplexing as to why this useful API requires workarounds when it could have been implemented internally.

**Comment by stephentoub** [comment]
We're talking past each other. Maybe someone else will be able to assist better.

**Comment by iSazonov** [comment]
@stephentoub Many thanks for your help! I get better understanding how this work internally - great experience for me.

**Comment by karelz** [comment]
Triage:
- We agree that `Stream.ReadTimeout` and `Stream.WriteTimeout` should NOT apply to async operations.
- We agree `HttpClient.Timeout` should NOT extend to stream read/write operations. It would break WebSockets, YARP and other customers.
- We agree that handling these timeouts only via cancellation is less than ideal and perhaps we should have something more user friendly exposed. E.g. something like `HttpClient.ReadWriteTimeout` or `HttpResponseStream.ReadWriteTimeout` (though we might not want to do that one) ... which may very well lead back to 1st case above (`Stream.ReadTimeout` or `Stream.WriteTimeout`).

@iSazonov is there anything else that you think needs attention / response from triage?

Given that there is way too long discussion here in many angles, we should fork the 3rd case above into separate issue and discuss solution there. Then close this issue. All assuming we didn't miss something else in the discussion.

**Comment by iSazonov** [comment]
> @iSazonov is there anything else that you think needs attention / response from triage?

No, 3rd case cover all. Who will open new issue? I feel we need to discuss timeouts for CopyToAsync() in common since it has a loop with base operations (ReadAsync and WriteAsync) which we can not control with CopyToAsync() interface as needed. Then the solution for HttpClient will be clear too.

**Comment by karelz** [comment]
@iSazonov why do you think we need to discuss `Stream.CopyToAsync` specifically?
I assume that once `ReadAsync` and `WriteAsync` are solved, that one should just built on them and therefore be addressed as well. Am I missing something?

I will file the issue.

**Comment by iSazonov** [comment]
@karelz We can use CTS with timeout for ReadAsync and WriteAsync in any scenario without problem - the timeout will terminate these methods. But if we use CST with timeout for CopyToAsync we can not control every ReadAsync and WriteAsync in the loop - the timeout will will only apply to the _whole_ cycle.

**Comment by karelz** [comment]
Created #75188. Closing.

**Comment by MihaZupan** [comment]
> ith timeout for CopyToAsync we can not control every ReadAsync and WriteAsync in the loop

Note that even if such a helper existed, it isn't what PowerShell should be using anyway if you are also tracking progress.
You are currently relying on `destination.Length`, but not all `Stream`s support querying the `Length` (as indicated by `CanSeek`).

**Comment by iSazonov** [comment]
> Note that even if such a helper existed, it isn't what PowerShell should be using anyway if you are also tracking progress.

This could help us to avoid the bug with hang we caught and make the code more reliable. (Progress bar is implemented in PowerSHell specific way in another thread.)

> You are currently relying on destination.Length, but not all Streams support querying the Length (as indicated by CanSeek).

The destination is always a file in PowerShell in the scenario.
